### PR TITLE
added a note about forking/threaded servers for deployment

### PR DESCRIPTION
**Pull request by jvanasco**
I saw the tweet about UWSGI recipes and felt compelled to submit this PR noting "special considerations" for forked/threaded servers (like uwsgi)

I'm sorry if the formatting is off, it's been a while since i've used REST.

### CODE REVIEW

File: docs/deployment/forked_threaded_servers.rst [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+Forked and Threaded Servers share common "gotchas" and solutions when
+using pyramid and some popular packages.
+
+Forked and Threaded servers tend to use a "copy on write" implementation detail
+to optimize how they work and share memory. This can create problems when
+certain actions happen before the fork or thread dispatch, such as when files or
+file-descriptors are opened or random number generators are initialized.
+
+Many servers have
```
Comment (bertjwregeer) [inline docs/deployment/forked_threaded_servers.rst]: Many servers have what?

File: docs/deployment/forked_threaded_servers.rst [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+========
+
+uwsgi offers a decorator to handle forking
+
+your application would include code like
+
+    from uwsgidecorators import postfork
+    
+    @postfork
+    def my_setuo():
```
Comment (bertjwregeer) [inline docs/deployment/forked_threaded_servers.rst]: s/setuo/setup/

File: docs/deployment/forked_threaded_servers.rst (hunk 1 above)
Comment (jvanasco) [reply docs/deployment/forked_threaded_servers.rst]: grrr. replaced from memory.  i tripped on a dog and spilled coffee on my keyboard this morning; switched to the backup keyboard and i've managed to accidentally cut/paste/delete/etc nonstop all day.

File: docs/deployment/deployment.rst [hunk 3]
```diff
@@ -52,3 +52,15 @@ A short list includes:
 + ``mod_wsgi``
 
 For more information on WSGI, see the `WSGI home <http://wsgi.readthedocs.io/en/latest/>`_
+
+
+Special Considerations
+++++++++++++++++++++++
+
+Certain environments and webservers require special considerations when
```
Comment (stevepiercy) [inline docs/deployment/deployment.rst]: Two words: "web servers"

File: docs/deployment/deployment.rst [hunk 4]
```diff
@@ -52,3 +52,15 @@ A short list includes:
...
+deploying your pyramid application due to implementation details of Python, the
```
Comment (stevepiercy) [inline docs/deployment/deployment.rst]: Capitalize "Pyramid"

File: docs/deployment/deployment.rst [hunk 5]
```diff
@@ -52,3 +52,15 @@ A short list includes:
...
+webserver, and/or popular packages.
```
Comment (stevepiercy) [inline docs/deployment/deployment.rst]: web server.

Do not use "and/or", "or" is less klunky and gets across the message.

File: docs/deployment/deployment.rst [hunk 6]
```diff
@@ -52,3 +52,15 @@ A short list includes:
...
+
+Forked and Threaded servers share some common gotchas and solutions.
+
++ ``forked_threaded_servers``
```
Comment (stevepiercy) [inline docs/deployment/deployment.rst]: Is this supposed to be a reference link? If so:

```
:doc:`forked_threaded_servers`
```

File: docs/deployment/forked_threaded_servers.rst [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+to optimize how they work and share memory. This can create problems when
+certain actions happen before the fork or thread dispatch, such as when files or
+file-descriptors are opened or random number generators are initialized.
+
+Many servers have built-in hooks or events which allow you to easily handle these
+situations.
+
+
+Servers
+++++++++++
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: Duplicate primary heading level violates rST syntax. The primary heading is the chapter title (the file itself), and secondary and lower headings are sections, sub-sections, etc. Use another underlining character.

Also underlines should match the length of the text that they underline.

File: docs/deployment/forked_threaded_servers.rst [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+gunicorn
+========
+
+gunicorn offers several hooks during an application lifecycle
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: Use proper sentence punctuation here and throughout: capitalize sentences, and end sentences with a period.

File: docs/deployment/forked_threaded_servers.rst [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+the postfork routine is provided as a function in a configuration python script.
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: The

File: docs/deployment/forked_threaded_servers.rst [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+for example a script 'config.py' might look like:
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: For

For names of files and inline code, use double backticks:

```
``config.py``
```

Finally we want to use syntax highlighting, which we should explicitly state with a `.. code-block::` directive, and not rely on defaults.

```
For example a script ``config.py`` might look like the following.

.. code_block:: python
```

File: docs/deployment/forked_threaded_servers.rst [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+	def post_fork(server, worker):
+		log.debug("gunicorn - post_fork")
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: Indents should be 4 spaces consistently.

File: docs/deployment/forked_threaded_servers.rst [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+and invoking the script would look like:
+
+	gunicorn --paste production.ini -c config.py
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]:
```
Invoking the script would look like the following.

.. code_block:: bash

    gunicorn --paste production.ini -c config.py
```

File: docs/deployment/forked_threaded_servers.rst [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+
+documentation for the `post_fork` hook appears on http://docs.gunicorn.org/en/latest/settings.html#post-fork
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]:
```
See `documentation for the post_fork hook <http://docs.gunicorn.org/en/latest/settings.html#post-fork>`_.
```

File: docs/deployment/forked_threaded_servers.rst [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+
+uwsgi
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: uWSGI, throughout

File: docs/deployment/forked_threaded_servers.rst [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+========
+
+uwsgi offers a decorator to handle forking
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: uWSGI and add period to end sentence.

File: docs/deployment/forked_threaded_servers.rst [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+
+your application would include code like
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: Your application would include code like the following.

```
.. code-block:: python

File: docs/deployment/forked_threaded_servers.rst [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+
+    from uwsgidecorators import postfork
+    
+    @postfork
+    def my_setup():
+		log.debug("uwsgi - postfork")
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: 4 spaces indent only.

File: docs/deployment/forked_threaded_servers.rst [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+
+documentation for the `postfork` decorator appears on https://uwsgi-docs.readthedocs.io/en/latest/PythonDecorators.html#uwsgidecorators.postfork
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: ```See `documentation for the postfork decorator <https://uwsgi-docs.readthedocs.io/en/latest/PythonDecorators.html#uwsgidecorators.postfork>`_.```

File: docs/deployment/forked_threaded_servers.rst [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+
+
+waitress
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: Waitress

File: docs/deployment/forked_threaded_servers.rst [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+========
+
+waitress is not a forking server, but it's threads can create similar issues as forking servers
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: Capitalization, "its", period.

Waitress is not a forking server, but its threads can create similar issues as forking servers.

File: docs/deployment/forked_threaded_servers.rst [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+
+
+Known Packages
+++++++++++++++
+
+
+SqlAlchemy
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: SQLAlchemy

File: docs/deployment/forked_threaded_servers.rst [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+============
+
+Many people use SqlAlchemy as part of their pyramid application stack.
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: SQLAlchemy, Pyramid

File: docs/deployment/forked_threaded_servers.rst [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+
+The database connections and the connection pools in SqlAlchemy are not safe to
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: SQLAlchemy

File: docs/deployment/forked_threaded_servers.rst [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+share across process boundaries (forks and threads). The connections and 
+connection pools are lazily created on their first use, so most pyramid users
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: Pyramid

File: docs/deployment/forked_threaded_servers.rst [hunk 25]
```diff
@@ -0,0 +1,94 @@
...
+
+Many people use SqlAlchemy as part of their pyramid application stack.
+
+The database connections and the connection pools in SqlAlchemy are not safe to
+share across process boundaries (forks and threads). The connections and 
+connection pools are lazily created on their first use, so most pyramid users 
+will not encounter an issue as database interaction usually happens on a 
+per-request basis.
+
+If your pyramid application connects to a database during the application startup
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: Pyramid

File: docs/deployment/forked_threaded_servers.rst [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+however, then you must use 'Engine.dispose' to reset the connections.  It might
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]:
```
``Engine.dispose``
```

File: docs/deployment/forked_threaded_servers.rst [hunk 27]
```diff
@@ -0,0 +1,94 @@
...
+look something like this:
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]:
```
look like the following.

.. code-block:: python
```

File: docs/deployment/forked_threaded_servers.rst [hunk 28]
```diff
@@ -0,0 +1,94 @@
...
+
+    @postfork
+    def reset_sqlalchemy():
+        models.engine.dispose()
+
+Additional documentation on this topic is available from SqlAlchemy's docs:
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: "SQLAlchemy's documentation."

File: docs/deployment/forked_threaded_servers.rst [hunk 29]
```diff
@@ -0,0 +1,94 @@
...
+
+* http://docs.sqlalchemy.org/en/latest/core/pooling.html#using-connection-pools-with-multiprocessing
+* http://docs.sqlalchemy.org/en/latest/core/connections.html#engine-disposal
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]:
```
* `Using Connection Pools with Multiprocessing <http://docs.sqlalchemy.org/en/latest/core/pooling.html#using-connection-pools-with-multiprocessing>`_
* `Engine Disposal <http://docs.sqlalchemy.org/en/latest/core/connections.html#engine-disposal>`_
```

File: docs/deployment/forked_threaded_servers.rst [hunk 30]
```diff
@@ -0,0 +1,94 @@
...
+
+
+PyCrypto
+============
+
+The pycrypto library provides for a `Crypto.Random.atfork` function to reseed the
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]:
```
The `PyCrypto <https://www.dlitz.net/software/pycrypto/>`_ library
```

File: docs/deployment/forked_threaded_servers.rst [hunk 31]
```diff
@@ -0,0 +1,94 @@
+Forked and Threaded Servers
++++++++++++++++++++++++++++
+
+Forked and Threaded Servers share common "gotchas" and solutions when
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: lowercase threaded servers

File: docs/deployment/forked_threaded_servers.rst [hunk 32]
```diff
@@ -0,0 +1,94 @@
...
+using pyramid and some popular packages.
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: Pyramid

File: docs/deployment/forked_threaded_servers.rst [hunk 33]
```diff
@@ -0,0 +1,94 @@
...
+
+Forked and Threaded servers tend to use a "copy on write" implementation detail
```
Comment (stevepiercy) [inline docs/deployment/forked_threaded_servers.rst]: lowercase threaded

File: docs/deployment/forked_threaded_servers.rst (hunk 13 above)
Comment (jvanasco) [reply docs/deployment/forked_threaded_servers.rst]: was the trailing underscore above intentional?

Comment (jvanasco) [reply docs/deployment/forked_threaded_servers.rst]: nevermind. i read the rst spec.

**Comment by jvanasco** [comment]
@stevepiercy thanks for the detailed feedback. i'm sorry i screwed up the formatting so much.

I've pushed the requested changes, but I also made a few small changes that need review:

1. I added a paragraph under the `Servers` and `Known Packages` headers
2. I slightly changed the wording in two random sentences.

Again, thank you for the detailed review and putting up with my formatting.  I think I subconsciously reverted to Markdown

**Comment by jvanasco** [comment]
fixed that stray typo.

rebase isn't happing cleanly, so i'm going to generate a patch + apply it to master.  i'll comment when that's done.

**Comment by jvanasco** [comment]
I reset this branch to the current master, applied the changes, and then did a `git push -f` to obliterate my previous commits.  I did stash the previous commits into a different branch though -- https://github.com/jvanasco/pyramid_cookbook/tree/feature-deployment_postfork-original

The now only commit is simply:

* updated authors file
* the updated paragraph in deployment, as fixed with the latest typo
* the previously approved forked_threaded_servers.rst file after the reviews above

**Pull request merged**